### ISSUE TITLE
Collection cache breaks on pagination!

### ISSUE THREAD

**Issue by alekseyl** [rails/rails#26330]
Hi everyone. just noticed that you added in 5.0 cache_key for collection, that's nice!
Quick search on repo gives that collection_cache relies on collection_cache_key
Right now it breaks with offset.

```
Entity.where(id: ids).limit(2).offset(1).cache_key
>>"entities/query-b730af29d2ff9ecd6fc4f9e853f7ec91-0"
Entity.where(id: ids).limit(2).offset(1).pluck(:id)
>>[2, 3]
Entity.find(2).touch
>>true
Entity.where(id: ids).limit(2).offset(1).cache_key
>> "entities/query-b730af29d2ff9ecd6fc4f9e853f7ec91-0"
```

I think this example is self-explained. 
the point break is: 

```
query = collection
          .unscope(:select)
          .select("COUNT(*) AS #{connection.quote_column_name("size")}", "MAX(#{column}) AS timestamp")
          .unscope(:order)
        result = connection.select_one(query)
```

It can be fixed by adding unscope on offset to query. 

`unscope(:offset)`

**PS**
ALSO It works in a very-very specific way and any paginated sub-collection will get different key if any element in a whole collection is touched. This will throw to garbage all sub-collection cache. 

 Couple month ago I also did cache_key on collections, **BUT I did it using PSQL** capabilities, since I don't use other DB-engines I didn't create general solution, but may be you would be interested in my solution and adapt it for different adapters :). Here it goes:

```
def cache_key( cache_columns = [:id, :updated_at] )
      # why use connection.execute instead of doing self.select because of an order. if you using some order on your scope
      # then columns you using to order must appear in the GROUP BY clause or be used in an aggregate function or you will get an error
      connection.
          execute( "SELECT md5(string_agg(#{cache_columns.map{|fld| "\"t\".\"#{fld}\"::text" }.join('||')}, '') ) as cache_key
                      FROM (#{ try(:to_sql) || all.to_sql }) t" )[0]['cache_key']
    end

```
 It's a part of a module which I include in ActiveRecord, so it works out of the box like this:

```
Model.cache_key
Model.some_scope.cache_key
```  

Works like charm, respecting order and without needing for re-cache untouched subcollection.

**Comment by alekseyl** [comment]
Double on https://github.com/rails/rails/issues/25454#issuecomment-243504057

But contains nice Postgres solution, which may be useful for those who using PG in their projects.

**Comment by maclover7** [comment]
Hi @alekseyl! As you already mentioned this issue is somewhat of a duplicate of #25454, and I believe there is already a PR in progress with a fix for this at #25496. Let's continue the discussion at #25454 to keep everything in one place. :)
